[LLMObs] add SDK updates (#23436)
DataDog/documentation: content/en/tracing/llm_observability/sdk.md

  api_key="<YOUR_DATADOG_API_KEY>",
  site="<YOUR_DATADOG_SITE>",
  agentless_enabled=True,
  integrations_enabled=True,
)
{{< /code-block >}}

`ml_app`
: optional - _string_
<br />The name of your LLM application, service, or project, under which all traces and spans are grouped. This helps distinguish between different applications or experiments. See [Application naming guidelines](#application-naming-guidelines) for allowed characters and other constraints. To override this value for a given trace, see [Tracing multiple applications](#tracing-multiple-applications). If not provided, this will default to the value of `DD_LLMOBS_APP_NAME`.

`integrations_enabled` - **default**: `true`
: optional - _boolean_ 

> <br />A flag to enable automatically tracing LLM calls for Datadog's supported [LLM integrations](#llm-integrations). If not provided, all supported LLM integrations will be enabled by default. To avoid using the LLM integrations, set this value to `false`.

Reviewer: ```suggestion
<br />A flag to enable automatically tracing LLM calls for Datadog's supported [LLM integrations](#llm-integrations). If not provided, all supported LLM integrations are enabled by default. To avoid using the LLM integrations, set this value to `false`.
```

: required - _string or numeric type_
<br />The value of the evaluation. Must be a string (for categorical `metric_type`) or integer/float (for score `metric_type`).

> `tags`

Reviewer 2: we only support `str:str` tags right now

The `ddtrace` library provides some out-of-the-box integrations that support distributed tracing for popular [web framework][11] and [HTTP][12] libraries. If your application makes requests using these supported libraries, you can enable distributed tracing by running:
{{< code-block lang="python">}}
from ddtrace import patch
patch(<INTEGRATION_NAME>=True)
{{< /code-block >}}

If your application does not use any of these supported libraries, you can enable distributed tracing by manually propagating span information to and from HTTP headers. The SDK provides the helper methods `LLMObs.inject_distributed_headers()` and `LLMObs.activate_distributed_headers()` to inject and activate tracing contexts in request headers.

#### Injecting distributed headers

The `LLMObs.inject_distributed_headers()` method takes a span and injects its context into the HTTP headers to be included in the request. This method accepts the following arguments:

`request_headers`
: required - _dictionary_
<br />The HTTP headers to extend with tracing context attributes.

`span`
: optional - _Span_ - **default**: `The current active span.`

> <br />The span to inject its context into the provided request headers. Any spans  (including those with function decorators), this defaults to the current active span.

Reviewer: ```suggestion
<br />The span to inject its context into the provided request headers. Any spans (including those with function decorators), this defaults to the current active span.
```

    return
{{< /code-block >}}

### Distributed tracing

The SDK supports tracing across distributed services or hosts.

> #### Automatic distributed tracing

Reviewer 2: Suggestion for the way this section should be phrased/structured

***
The SDK supports tracing across distributed services or hosts. Distributed tracing works by propagating span information across web requests.

If your application makes requests using these supported web framework and http libraries, you can enable distributed tracing by running:
ddtrace.patch(<INTEGRATION_NAME>)

If your application does not use any of these supported libraries, you can enable distributed tracing by manually propagating span information from http headers.
.... 
***

Reviewer 2: Mainly because this clarifies what is exactly being "manually" vs "automatically" done (injecting and extracting from request headers)

> The `ddtrace` library provides some out-of-the-box integrations that support distributed tracing for popular [web framework][11] and [HTTP][12] libraries. If you are using any of these libraries, you must enable the corresponding integration via the `ddtrace.patch(<INTEGRATION_NAME>)` method. For example if you are using the `urllib3` and `fastAPI` libraries, you should add these to the top of your application's entrypoint file:

Reviewer: ```suggestion
The `ddtrace` library provides some out-of-the-box integrations that support distributed tracing for popular [web framework][11] and [HTTP][12] libraries. If you are using any of these libraries, enable the corresponding integration through the `ddtrace.patch(<INTEGRATION_NAME>)` method. For example, if you are using the `urllib3` and `fastAPI` libraries, add these to the top of your application's entrypoint file:
```

{{< code-block lang="python">}}
from ddtrace import patch
patch(urllib3=True, fastapi=True)
{{< /code-block >}}

#### Manual distributed tracing

If you are using libraries that are not supported by the `ddtrace` library's integrations, the SDK provides two helper methods `LLMObs.inject_distributed_headers()` and `LLMObs.activate_distributed_headers()` to manually inject and propagate tracing contexts in distributed request headers.

##### Injecting distributed headers

The `LLMObs.inject_distributed_headers()` method takes a span and injects its context into the HTTP headers to be included in the request. This method accepts the following arguments:

`request_headers`
: required - _dictionary_
<br />The HTTP headers to extend with tracing context attributes.

`span`
: optional - _Span_ - **default**: `The current active span.`

> <br />The span to inject its context into the provided request headers. If not provided (as when using function decorators), this will default to the current active span.

Reviewer: ```suggestion
<br />The span to inject its context into the provided request headers. If not provided (as when using function decorators), this defaults to the current active span.
```

##### Activating distributed headers

The `LLMObs.activate_distributed_headers()` method takes HTTP headers and extracts tracing context attributes to activate in the new service.

> **Note**: You must call `LLMObs.activate_distributed_headers()` before starting any spans in your downstream service. Any spans started prior (including via fucntion decorators) will not be captured in the distributed trace.

Reviewer: ```suggestion
**Note**: You must call `LLMObs.activate_distributed_headers()` before starting any spans in your downstream service. Any spans started prior (including those with function decorators) are not captured in the distributed trace.
```
